### PR TITLE
219 scales parameter for facet_wrap in g_spaghettiplot

### PR DESCRIPTION
**Pull request by m7pr**
Fix for https://github.com/insightsengineering/teal.goshawk/issues/219 and continued here https://github.com/insightsengineering/teal.goshawk/pull/227

# Data
```{R}
library(stringr)

# original ARM value = dose value
arm_mapping <- list(
  "A: Drug X" = "150mg QD", "B: Placebo" = "Placebo", "C: Combination" = "Combination"
)
color_manual <- c("150mg QD" = "#000000", "Placebo" = "#3498DB", "Combination" = "#E74C3C")

ADLB <- goshawk::rADLB
var_labels <- lapply(ADLB, function(x) attributes(x)$label)
ADLB <- ADLB %>%
  mutate(AVISITCD = case_when(
    AVISIT == "SCREENING" ~ "SCR",
    AVISIT == "BASELINE" ~ "BL",
    grepl("WEEK", AVISIT) ~
      paste(
        "W",
        trimws(
          substr(
            AVISIT,
            start = 6,
            stop = str_locate(AVISIT, "DAY") - 1
          )
        )
      ),
    TRUE ~ NA_character_
  )) %>%
  mutate(AVISITCDN = case_when(
    AVISITCD == "SCR" ~ -2,
    AVISITCD == "BL" ~ 0,
    grepl("W", AVISITCD) ~ as.numeric(gsub("\\D+", "", AVISITCD)),
    TRUE ~ NA_real_
  )) %>%
  # use ARMCD values to order treatment in visualization legend
  mutate(TRTORD = ifelse(grepl("C", ARMCD), 1,
    ifelse(grepl("B", ARMCD), 2,
      ifelse(grepl("A", ARMCD), 3, NA)
    )
  )) %>%
  mutate(ARM = as.character(arm_mapping[match(ARM, names(arm_mapping))])) %>%
  mutate(ARM = factor(ARM) %>%
    reorder(TRTORD)) %>%
  mutate(ANRLO = .5, ANRHI = 1) %>%
  rowwise() %>%
  group_by(PARAMCD) %>%
  mutate(LBSTRESC = ifelse(USUBJID %in% sample(USUBJID, 1, replace = TRUE),
    paste("<", round(runif(1, min = .5, max = .7))), LBSTRESC
  )) %>%
  mutate(LBSTRESC = ifelse(USUBJID %in% sample(USUBJID, 1, replace = TRUE),
    paste(">", round(runif(1, min = .9, max = 1.2))), LBSTRESC
  )) %>%
  ungroup()
attr(ADLB[["ARM"]], "label") <- var_labels[["ARM"]]
attr(ADLB[["ANRLO"]], "label") <- "Analysis Normal Range Lower Limit"
attr(ADLB[["ANRHI"]], "label") <- "Analysis Normal Range Upper Limit"

# add LLOQ and ULOQ variables
ADLB_LOQS <- goshawk:::h_identify_loq_values(ADLB)
ADLB <- left_join(ADLB, ADLB_LOQS, by = "PARAM")
```

# Show all scales - `facet_scales = 'fixed'`

```{R}
g_spaghettiplot(
  data = ADLB,
  subj_id = "USUBJID",
  biomarker_var = "PARAMCD",
  biomarker = "CRP",
  value_var = "AVAL",
  trt_group = "ARM",
  time = "RACE",
  color_manual = color_manual,
  color_comb = "#39ff14",
  alpha = .02,
  facet_scales = "fixed",
  rotate_xlab = FALSE,
  group_stats = "median",
  hline_arb = c(.5, .7, 1),
  hline_arb_color = c("blue", "red", "green"),
  hline_arb_label = c("Arb_Hori_line_A", "Arb_Hori_line_B", "Arb_Hori_line_C"),
  hline_vars = c("ANRHI", "ANRLO")
)
```
image

## Free scales for x - `facet_scales = 'free_x'`

```{R}
g_spaghettiplot(
  data = ADLB,
  subj_id = "USUBJID",
  biomarker_var = "PARAMCD",
  biomarker = "CRP",
  value_var = "AVAL",
  trt_group = "ARM",
  time = "RACE",
  color_manual = color_manual,
  color_comb = "#39ff14",
  alpha = .02,
  facet_scales = "free_x",
  rotate_xlab = FALSE,
  group_stats = "median",
  hline_arb = c(.5, .7, 1),
  hline_arb_color = c("blue", "red", "green"),
  hline_arb_label = c("Arb_Hori_line_A", "Arb_Hori_line_B", "Arb_Hori_line_C"),
  hline_vars = c("ANRHI", "ANRLO")
)
```

image

### CODE REVIEW

File: R/g_spaghettiplot.R [hunk 1]
```diff
@@ -182,6 +184,7 @@ g_spaghettiplot <- function(data,
                             ylim = c(NA, NA),
                             alpha = 1.0,
                             facet_ncol = 2,
+                            facet_scales = "fixed",
```
Comment (chlebowa) [inline R/g_spaghettiplot.R:187]: add `match.arg` for allowed values

**Review by chlebowa — CHANGES_REQUESTED** [review]

File: R/g_spaghettiplot.R [hunk 2]
```diff
@@ -21,6 +21,8 @@
 #'   if the default limits are not suitable.
 #' @param alpha subject line transparency (0 = transparent, 1 = opaque)
 #' @param facet_ncol number of facets per row.
+#' @param facet_scales passed to \code{\link[ggplot2]{facet_wrap}} \code{scales} parameter. Should scales be fixed (`"fixed"`,
```
Comment (chlebowa) [inline R/g_spaghettiplot.R:24]:
```suggestion
#' @param facet_scales passed to `scales` in [`ggplot2::facet_wrap`]. Should scales be fixed (`"fixed"`,
```

File: NEWS.md [hunk 3]
```diff
@@ -14,6 +14,7 @@
 ### Enhancements
 
 * Implemented `nestcolor` with slight refactoring to `g_lineplot`, `g_density_distribution_plot` and added `nestcolor` in examples without custom color manuals.
+* `g_spaghettiplot` has new `facet_scales` parameter that allow to free the scales for missing factor levels within a piece of the grid in the final plot.
```
Comment (chlebowa) [inline NEWS.md:17]:
```suggestion
* `g_spaghettiplot` has a new parameter `facet_scales` that allows for releasing scales in plot facets.
```

File: R/g_spaghettiplot.R (hunk 1 above)
Comment (m7pr) [reply R/g_spaghettiplot.R]: yup! ab67cd6442270ac5f62fee34c89cc7e6614757b3

Comment (m7pr) [reply R/g_spaghettiplot.R]: this is checked anyway in ggplot2::facet_wrap

> scales <- arg_match0(scales %||% "fixed", c("fixed", "free_x", "free_y", "free"))

Comment (chlebowa) [reply R/g_spaghettiplot.R]: And will be checked again in the module, I suppose :)
When you add the argument and an input to the module, remember to adjust the `InputValidator`.

**Review by chlebowa — APPROVED** [review]

**Pull request merged**